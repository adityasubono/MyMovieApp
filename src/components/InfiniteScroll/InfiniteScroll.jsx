import React, {useRef, useCallback} from "react";

import MovieItem from './Item/MovieItem'

const InfiniteScroll = ({
                            hasMoreData,
                            listData = [],
                            isLoading,
                            onBottomHit,
                            onImageClick,
                            onBodyClick
                        }) => {
    const observer = useRef();

    const lastBookElementRef = useCallback(
        (node) => {
            if (isLoading) {
                return;
            }

            /**
             * This function will look for the last element from list data
             * and will fetch the next page data if element found
             */
            if (observer.current) {
                observer.current.disconnect();
            }

            observer.current = new IntersectionObserver((entries) => {
                if (entries[0].isIntersecting && hasMoreData && onBottomHit) {
                    onBottomHit()
                }
            });

            if (node) {
                observer.current.observe(node);
            }
        },
        [isLoading, hasMoreData, onBottomHit]
    );

    return (
        <>
            <div className="row justify-content-center" data-testid="infinitescroll-test">
                {listData.map((data, idx) => <MovieItem key={idx}
                                                        item={data} idx={idx}
                                                        elementRef={lastBookElementRef} records={listData.length}
                                                        onImageClick={onImageClick} onBodyClick={onBodyClick}
                />)}
            </div>
            {isLoading && <div className="row justify-content-center align-items-center" style={{height: '150px'}}>
                <div className="col-md-2 d-flex justify-content-center" data-testid="loading-test">
                    <p>{"Loading..."}</p>
                </div>
            </div>}
        </>
    );
}


export default InfiniteScroll;
