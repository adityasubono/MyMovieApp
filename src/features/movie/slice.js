import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  movies: [],
  totalResults: 0,
  success: false,
  movie: {}
};

const movieSlice = createSlice({
    name: 'movie',
    initialState,
    reducers: {
      getListSuccess: (state, action) => {
        state.success = true;
        state.movies = (action.payload.movies && 
          action.payload.movies.map((item) => ({ 
            ...item, Poster: item.Poster !== 'N/A' ? item.Poster : null
          }))) || [];

        state.totalResults = action.payload.totalResults;
      },
      getListError: (state, action) => {
        state.movies = [];
        state.totalResults = 0;
        state.success = false;
      },
      detailSuccess: (state, action) => {
        const movie = action.payload.movie;

        state.success = true;
        state.movie = movie;
        state.movie.Actors = movie.Actors.split(', ') || [];
        state.movie.Genre = movie.Genre.split(', ') || [];
        state.movie.Poster = movie.Poster !== 'N/A' ? movie.Poster : null;
      },
      detailLoading: (state) => {
        state.success = false;
        state.movie = {};
      }
    },
  });

export const { getListSuccess, getListError, detailSuccess, detailLoading } = movieSlice.actions;

export default movieSlice.reducer;
