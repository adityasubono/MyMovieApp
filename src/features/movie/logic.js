import { createAsyncThunk } from '@reduxjs/toolkit';
import axiosInstance from '../../app/axiosConfig';

import { getListSuccess, detailSuccess, getListError, detailLoading } from './slice';

export const getList = createAsyncThunk(
    'movie/fetchListMovie',
    async ({ searchTerm, page }, thunkAPI) => {
      const response = await axiosInstance(`/?s=${searchTerm}&page=${page}`);

      if (response.data.Response === "False") {
        thunkAPI.dispatch(getListError());
        return response.data;
      }
      
      thunkAPI.dispatch(getListSuccess({ movies: response.data.Search, totalResults: (response.data.totalResults && parseInt(response.data.totalResults)) || 0 }));
      return response.data;
    }
  );


export const detailMovieById = createAsyncThunk(
  'movie/fetchDetailMovie',
  async (id, thunkAPI) => {
    thunkAPI.dispatch(detailLoading());
    const response = await axiosInstance(`/?i=${id}`);
    
    thunkAPI.dispatch(detailSuccess({ movie: response.data }));
    return response.data;
  }
);