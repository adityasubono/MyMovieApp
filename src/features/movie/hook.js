import { useEffect, useState, useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {getList, detailMovieById} from './logic';


export const useMovies = () => {
    const dispatch = useDispatch();
    const moviesRaw = useSelector((state) => state.movie.movies);
    const isSuccess = useSelector((state) => state.movie.success);

    const [searchTerm, setSearchTerm] = useState('');
    const [page, setPage] = useState(1);
    const [loading, setLoading] = useState(false);
    const [movies, setMovies] = useState([]);

    useEffect(() => {
        /**
        * This function will fetch data from the API
        * with given parameter from the modules
        */
        setLoading(true);
        const dispatcher = dispatch(getList({ searchTerm, page }));

        dispatcher.then(() => {
            setLoading(false);
        });

        return () => {
            dispatcher.abort();
            setLoading(false);
        };
    }, [searchTerm]);

    useEffect(() => {
        setMovies([]);
    }, [searchTerm]);

    useEffect(() => {
        /**
         * This function will create a new array with existing list data and
         * additional data from the API based on pagination and wil return array with unique item
         */
         if (!isSuccess) {
            setMovies([]);
         } else {
             const key = 'imdbID';
             setMovies((prev) => {
                 return  [...new Map([...prev, ...moviesRaw].map(item =>
                     [item[key], item])).values()];
               });
         }
    }, [page, searchTerm, moviesRaw]);

    const onLoadMore = useCallback(async () => {
        setLoading(true);
        await dispatch(getList({ searchTerm, page: page + 1 }));
        setLoading(false);
        setPage(page + 1);
    }, [page, searchTerm]);

    return {loading, movies, searchTerm, setSearchTerm, onLoadMore, setPage};
}

export const useMovie = (id) => {
    const dispatch = useDispatch();
    const movie = useSelector((state) => state.movie.movie);

    const [loading, setLoading] = useState(false);

    useEffect(() => {
        /**
        * This function will fetch data from the API
        * with given parameter from the modules
        */
        setLoading(true);
        const dispatcher = dispatch(detailMovieById(id));

        dispatcher.then(() => {
            setLoading(false);
        });

        return () => {
            dispatcher.abort();
            setLoading(false);
        };
    }, []);

    return {movie, loading}
}
