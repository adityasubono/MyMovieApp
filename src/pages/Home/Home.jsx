import React, { useCallback, useState } from 'react';
import {useHistory} from 'react-router-dom';

import { useMovies } from '../../features/movie/hook';
import { debounce } from '../../utils/utils';

import InfiniteScrolls from '../../components/InfiniteScroll/InfiniteScroll';
import Modal from '../../components/Modal/Modal';

import '../../App.css';
import Loading from '../../assets/loading.gif';

const Apps = () => {
  const [visible, setVisible] = useState(false);
  const [selectedMovie, setMovie] = useState(null);

  const history = useHistory();
  const { movies, setSearchTerm, loading, onLoadMore, setPage } = useMovies();

  const hasMoreData = movies.length < 1000;

  const openDetail = useCallback((id) => () =>{
    history.push('/detail/' + id);
  }, [history]);

  const toggleModal = useCallback(() => {
    setVisible(!visible);
  }, [visible])

  const showImage = useCallback((movie) => () => {
      /**
       * function to open the modal from each element data
       */
    toggleModal();
    setMovie(movie)
  }, [toggleModal])

  const onQueryChange = useCallback(debounce((e) => {
      /**
       * function to chanfe the search term based on user input
       * will reset the page if there is new value omitted
       */
    setSearchTerm(e.target.value);
    setPage(1);
  }, 1000), []);


  return (
    <div className="row">


      <section className="py-5 text-center container">
        <div className="row py-lg-5">
          <div className="col-lg-6 col-md-8 mx-auto">
            <h1 className="fw-light">Kami Hadir Untuk Anda</h1>
            <p className="lead text-muted">Kebosanan pasti ada ya... , Nah untuk menghilangan kebosaan apalagi anda
              banyak disibukan bekerja dirumah selama pandemi COVIC-19 ini dan pasti jenuh sekali. Nah MyMovie.com hadir
              untuk membunuh rasa bosan anda dan ditemani film favorit anda, yuk segara cari film kesayangan anda disini</p>
            <div className="col-md-12 d-flex justify-content-center p-3">
              <input
                  type="text"
                  className="form-control"
                  placeholder="Yuk Cari Film Anda Disini..."
                  style={{ maxWidth: '50%', borderWidth: 2, borderColor: 'grey', textAlign: 'center'}}
                  onChange={onQueryChange}
              />
            </div>
          </div>
        </div>
      </section>



      <InfiniteScrolls hasMoreData={hasMoreData}
        isLoading={loading}
        onBottomHit={onLoadMore}
        listData={movies}
        onImageClick={showImage}
        onBodyClick={openDetail} />

      {selectedMovie && <Modal
        visible={visible}
        toggle={toggleModal}
        title={`${selectedMovie.Title || ''} (${selectedMovie.Year})`}
        content={<div className="row">
            <div className="col-md-12 d-flex justify-content-center">
              <img src={((selectedMovie.Poster !== 'N/A') && selectedMovie.Poster) || Loading} alt=""/>
            </div>
          </div>
        }
      />}
    </div>

  )
}

export default Apps;
