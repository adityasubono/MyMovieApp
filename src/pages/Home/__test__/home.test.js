import { cleanup, render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { shallow } from 'enzyme';

import  * as hooks from '../../../features/movie/hook';

import Home from '../Home';
import InfiniteScroll from '../../../components/InfiniteScroll/InfiniteScroll';

const dummyData = {
    "Search": [
      {
        "Title": "Batman: Arkham Knight",
        "Year": "2015",
        "imdbID": "tt3554580",
        "Type": "game",
        "Poster": "https://m.media-amazon.com/images/M/MV5BMTc0MTcxMzQ0Ml5BMl5BanBnXkFtZTgwNDc3MzE0MTE@._V1_SX300.jpg"
      },
      {
        "Title": "Batman: The Brave and the Bold",
        "Year": "2008–2011",
        "imdbID": "tt1213218",
        "Type": "series",
        "Poster": "https://m.media-amazon.com/images/M/MV5BMDBiODMwZmYtYWFiZC00N2I2LWJiMDktMGYwM2QxNzUxMzc3XkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg"
      },
      {
        "Title": "Batman: Dead End",
        "Year": "2003",
        "imdbID": "tt0374526",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BMTkzMDI0MTIyNF5BMl5BanBnXkFtZTYwMTUxMTY3._V1_SX300.jpg"
      },
      {
        "Title": "Batman Beyond: The Movie",
        "Year": "1999",
        "imdbID": "tt0231237",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BNTJjNmViYjEtNGRjZS00MDk3LThmZjktNzhkZDg1YjBiNjA0XkEyXkFqcGdeQXVyNjYyODY4NDU@._V1_SX300.jpg"
      },
      {
        "Title": "Batman: Return of the Caped Crusaders",
        "Year": "2016",
        "imdbID": "tt5973626",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BOTdlNjNhOGEtZTk2Yy00MjI5LTllOTMtODczMWFmNmNhZmVlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg"
      },
      {
        "Title": "Lego Batman: The Movie - DC Super Heroes Unite",
        "Year": "2013",
        "imdbID": "tt2465238",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BNjRjOGU2NzUtMjAwOC00MDI3LThmNmUtNTVkZTQ0MDEyYTc1XkEyXkFqcGdeQXVyMTA5NzUzODM4._V1_SX300.jpg"
      },
      {
        "Title": "Beware the Batman",
        "Year": "2013–2014",
        "imdbID": "tt2329077",
        "Type": "series",
        "Poster": "https://m.media-amazon.com/images/M/MV5BY2IzMGUzN2UtZDI4NS00NWM3LTgyNTYtYWJiNTg5OGNjNWQ0XkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg"
      },
      {
        "Title": "Batman: Soul of the Dragon",
        "Year": "2021",
        "imdbID": "tt12885852",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BYTUxNDQ3OGMtNWM2Yy00MDEwLThkMmUtZGVmZWM5MTczNjExXkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg"
      },
      {
        "Title": "Batman vs. Two-Face",
        "Year": "2017",
        "imdbID": "tt6142314",
        "Type": "movie",
        "Poster": "https://m.media-amazon.com/images/M/MV5BY2MyY2NjNzgtZDAxNi00OWM2LTg3MDUtOTcyOGU1OTdjODNjXkEyXkFqcGdeQXVyMTg2NjYzOA@@._V1_SX300.jpg"
      },
      {
        "Title": "Batman: The Telltale Series",
        "Year": "2016",
        "imdbID": "tt5785964",
        "Type": "game",
        "Poster": "https://m.media-amazon.com/images/M/MV5BMDQ0YTU2ZjgtZmZiZC00YWM2LTljYWYtOWI2NzUzZmIwMzg4XkEyXkFqcGdeQXVyNDQ2MTMzODA@._V1_SX300.jpg"
      }
    ],
    "totalResults": "444",
    "Response": "True"
  }

describe('Home Component Test', () => {
    afterEach(cleanup)
    beforeEach(() => {
        jest.spyOn(hooks, 'useMovies').mockImplementation(() => ({movies: dummyData}));
    })
    it('should render home component and match with snapshot', () => {
        const wrapper = shallow(<Home />);
        expect(wrapper.getElements()).toMatchSnapshot();
    });

    it('should render infinite scroll component in home component', () => {
        const wrapper = shallow(<Home />);
        expect(wrapper.containsMatchingElement(<InfiniteScroll />)).toBeTruthy();
    });
})