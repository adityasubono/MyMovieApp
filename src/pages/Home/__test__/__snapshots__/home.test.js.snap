// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Component Test should render home component and match with snapshot 1`] = `
Array [
  <div
    className="row"
  >
    <div
      className="col-md-12 d-flex justify-content-center p-3"
    >
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Input movie title here..."
        style={
          Object {
            "maxWidth": "50%",
          }
        }
        type="text"
      />
    </div>
    <InfiniteScroll
      hasMoreData={false}
      listData={
        Object {
          "Response": "True",
          "Search": Array [
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BMTc0MTcxMzQ0Ml5BMl5BanBnXkFtZTgwNDc3MzE0MTE@._V1_SX300.jpg",
              "Title": "Batman: Arkham Knight",
              "Type": "game",
              "Year": "2015",
              "imdbID": "tt3554580",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BMDBiODMwZmYtYWFiZC00N2I2LWJiMDktMGYwM2QxNzUxMzc3XkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg",
              "Title": "Batman: The Brave and the Bold",
              "Type": "series",
              "Year": "2008–2011",
              "imdbID": "tt1213218",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BMTkzMDI0MTIyNF5BMl5BanBnXkFtZTYwMTUxMTY3._V1_SX300.jpg",
              "Title": "Batman: Dead End",
              "Type": "movie",
              "Year": "2003",
              "imdbID": "tt0374526",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BNTJjNmViYjEtNGRjZS00MDk3LThmZjktNzhkZDg1YjBiNjA0XkEyXkFqcGdeQXVyNjYyODY4NDU@._V1_SX300.jpg",
              "Title": "Batman Beyond: The Movie",
              "Type": "movie",
              "Year": "1999",
              "imdbID": "tt0231237",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BOTdlNjNhOGEtZTk2Yy00MjI5LTllOTMtODczMWFmNmNhZmVlXkEyXkFqcGdeQXVyNTAyODkwOQ@@._V1_SX300.jpg",
              "Title": "Batman: Return of the Caped Crusaders",
              "Type": "movie",
              "Year": "2016",
              "imdbID": "tt5973626",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BNjRjOGU2NzUtMjAwOC00MDI3LThmNmUtNTVkZTQ0MDEyYTc1XkEyXkFqcGdeQXVyMTA5NzUzODM4._V1_SX300.jpg",
              "Title": "Lego Batman: The Movie - DC Super Heroes Unite",
              "Type": "movie",
              "Year": "2013",
              "imdbID": "tt2465238",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BY2IzMGUzN2UtZDI4NS00NWM3LTgyNTYtYWJiNTg5OGNjNWQ0XkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg",
              "Title": "Beware the Batman",
              "Type": "series",
              "Year": "2013–2014",
              "imdbID": "tt2329077",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BYTUxNDQ3OGMtNWM2Yy00MDEwLThkMmUtZGVmZWM5MTczNjExXkEyXkFqcGdeQXVyMTEyNzgwMDUw._V1_SX300.jpg",
              "Title": "Batman: Soul of the Dragon",
              "Type": "movie",
              "Year": "2021",
              "imdbID": "tt12885852",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BY2MyY2NjNzgtZDAxNi00OWM2LTg3MDUtOTcyOGU1OTdjODNjXkEyXkFqcGdeQXVyMTg2NjYzOA@@._V1_SX300.jpg",
              "Title": "Batman vs. Two-Face",
              "Type": "movie",
              "Year": "2017",
              "imdbID": "tt6142314",
            },
            Object {
              "Poster": "https://m.media-amazon.com/images/M/MV5BMDQ0YTU2ZjgtZmZiZC00YWM2LTljYWYtOWI2NzUzZmIwMzg4XkEyXkFqcGdeQXVyNDQ2MTMzODA@._V1_SX300.jpg",
              "Title": "Batman: The Telltale Series",
              "Type": "game",
              "Year": "2016",
              "imdbID": "tt5785964",
            },
          ],
          "totalResults": "444",
        }
      }
      onBodyClick={[Function]}
      onImageClick={[Function]}
    />
  </div>,
]
`;
